File: src/graph/Caret.js
function parsegraph_Caret(nodeRoot, font)
{
    if(arguments.length === 0) {
        nodeRoot = new parsegraph_Node(parsegraph_DEFAULT_NODE_TYPE);
    }
    if(typeof nodeRoot != "object") {
        nodeRoot = new parsegraph_Node(parsegraph_readNodeType(nodeRoot));
    }
    this._nodeRoot = nodeRoot;

    this._mathMode = false;

    // Stack of nodes.
    this._nodes = [this._nodeRoot];

    // A mapping of nodes to their saved names.
    this._savedNodes = null;

    this._labels = [];

    this._font = font ? font : parsegraph_defaultFont();
    this._world = null;
};

parsegraph_Caret.prototype.clone = function()
{
    return new parsegraph_Caret(this.node(), this.font());
};

parsegraph_Caret.prototype.setMathMode = function(mathMode)
{
    this._mathMode = mathMode;
    var curr = this.node();
    if(mathMode) {
        switch(curr.type()) {
        case parsegraph_BLOCK:
            curr.setBlockStyle(parsegraph_BLOCK_MATH_STYLE);
            break;
        case parsegraph_SLOT:
            curr.setBlockStyle(parsegraph_SLOT_MATH_STYLE);
            break;
        }
    }
};

parsegraph_Caret.prototype.mathMode = function()
{
    return this._mathMode;
};

parsegraph_Caret_Tests = new parsegraph_TestSuite("parsegraph_Caret");
parsegraph_Caret_Tests.addTest("new parsegraph_Caret", function() {
    var car = new parsegraph_Caret('s');
    var n = new parsegraph_Node(parsegraph_BLOCK);
    car = new parsegraph_Caret(n);
    car = new parsegraph_Caret();
    if(car.node().type() !== parsegraph_DEFAULT_NODE_TYPE) {
        console.log(parsegraph_DEFAULT_NODE_TYPE);
        return car.node().type() + " is not the default.";
    }
});

parsegraph_Caret.prototype.setFont = function(font)
{
    this._font = font;
};

parsegraph_Caret.prototype.font = function()
{
    if(!this._font) {
        throw new Error("Caret does not have a Font");
    }
    return this._font;
};

parsegraph_Caret.prototype.node = function()
{
    if(this._nodes.length === 0) {
        throw parsegraph_createException(parsegraph_NO_NODE_FOUND);
    }
    return this._nodes[this._nodes.length - 1];
};

parsegraph_Caret.prototype.has = function(inDirection)
{
    inDirection = parsegraph_readNodeDirection(inDirection);
    return this.node().hasNode(inDirection);
};

parsegraph_Caret.prototype.spawn = function(inDirection, newType, newAlignmentMode)
{
    // Interpret the given direction and type for ease-of-use.
    inDirection = parsegraph_readNodeDirection(inDirection);
    newType = parsegraph_readNodeType(newType);

    // Spawn a node in the given direction.
    var created = this.node().spawnNode(inDirection, newType);

    // Use the given alignment mode.
    if(newAlignmentMode !== undefined) {
        newAlignmentMode = parsegraph_readNodeAlignment(newAlignmentMode);
        this.align(inDirection, newAlignmentMode);
        if(newAlignmentMode !== parsegraph_DO_NOT_ALIGN) {
            this.node().setNodeFit(parsegraph_NODE_FIT_EXACT);
        }
    }

    if(this._mathMode) {
        switch(newType) {
        case parsegraph_BLOCK:
            created.setBlockStyle(parsegraph_BLOCK_MATH_STYLE);
            break;
        case parsegraph_SLOT:
            created.setBlockStyle(parsegraph_SLOT_MATH_STYLE);
            break;
        }
    }

    return created;
};

parsegraph_Caret.prototype.connect = function(inDirection, node)
{
    // Interpret the given direction for ease-of-use.
    inDirection = parsegraph_readNodeDirection(inDirection);

    this.node().connectNode(inDirection, node);

    return node;
};

parsegraph_Caret.prototype.disconnect = function(inDirection)
{
    if(arguments.length > 0) {
        // Interpret the given direction for ease-of-use.
        inDirection = parsegraph_readNodeDirection(inDirection);
        return this.node().disconnectNode(inDirection);
    }

    if(this.node().isRoot()) {
        return this.node();
    }

    return this.node().parentNode().disconnectNode(parsegraph_reverseNodeDirection(this.node().parentDirection()));
};

parsegraph_Caret.prototype.crease = function(inDirection)
{
    // Interpret the given direction for ease-of-use.
    inDirection = parsegraph_readNodeDirection(inDirection);

    var node;
    if(arguments.length === 0) {
        node = this.node();
    }
    else {
        node = this.node().nodeAt(inDirection);
    }

    // Create a new paint group for the connection.
    if(!node.localPaintGroup()) {
        node.setPaintGroup(true);
    }
};

parsegraph_Caret.prototype.setWorld = function(world)
{
    this._world = world;
};

parsegraph_Caret.prototype.freeze = function(inDirection)
{
    // Interpret the given direction for ease-of-use.
    inDirection = parsegraph_readNodeDirection(inDirection);
    var node;
    if(arguments.length === 0) {
        node = this.node();
    }
    else {
        node = this.node().nodeAt(inDirection);
    }
    if(!this._world) {
        throw new Error("Caret must have a world in order to freeze nodes");
    }
    node.freeze(this._world.freezer());
};

parsegraph_Caret.prototype.thaw = function(inDirection)
{
    // Interpret the given direction for ease-of-use.
    inDirection = parsegraph_readNodeDirection(inDirection);
    var node;
    if(arguments.length === 0) {
        node = this.node();
    }
    else {
        node = this.node().nodeAt(inDirection);
    }
    node.thaw();
};

parsegraph_Caret.prototype.uncrease = function(inDirection)
{
    // Interpret the given direction for ease-of-use.
    inDirection = parsegraph_readNodeDirection(inDirection);

    var node;
    if(arguments.length === 0) {
        node = this.node();
    }
    else {
        node = this.node().nodeAt(inDirection);
    }

    // Remove the paint group.
    node.setPaintGroup(false);
};

parsegraph_Caret.prototype.isCreased = function(inDirection)
{
    // Interpret the given direction for ease-of-use.
    inDirection = parsegraph_readNodeDirection(inDirection);

    var node;
    if(arguments.length === 0) {
        node = this.node();
    }
    else {
        node = this.node().nodeAt(inDirection);
    }

    return !!node.localPaintGroup();
};
parsegraph_Caret.prototype.creased = parsegraph_Caret.prototype.isCreased;

parsegraph_Caret.prototype.erase = function(inDirection)
{
    inDirection = parsegraph_readNodeDirection(inDirection);
    return this.node().eraseNode(inDirection);
};

parsegraph_Caret.prototype.onClick = function(clickListener, thisArg)
{
    this.node().setClickListener(clickListener, thisArg);
};

parsegraph_Caret.prototype.onChange = function(changeListener, thisArg)
{
    this.node().setChangeListener(changeListener, thisArg);
};

parsegraph_Caret.prototype.onKey = function(keyListener, thisArg)
{
    this.node().setKeyListener(keyListener, thisArg);
};

parsegraph_Caret_Tests.addTest("parsegraph_Caret.onKey", function() {
    var car = new parsegraph_Caret();
    car.onKey(function() {
        console.log("Key pressed");
    });
});

parsegraph_Caret.prototype.move = function(toDirection)
{
    toDirection = parsegraph_readNodeDirection(toDirection);
    var dest = this.node().nodeAt(toDirection);
    if(!dest) {
        throw parsegraph_createException(parsegraph_NO_NODE_FOUND);
    }
    this._nodes[this._nodes.length - 1] = dest;
};

parsegraph_Caret.prototype.push = function()
{
    this._nodes.push(this.node());
};

parsegraph_Caret.prototype.save = function(id)
{
    if(id === undefined) {
        id = parsegraph_generateID();
    }
    if(!this._savedNodes) {
        this._savedNodes = {};
    }
    this._savedNodes[id] = this.node();
    return id;
}

parsegraph_Caret.prototype.clearSave = function(id)
{
    if(!this._savedNodes) {
        return;
    }
    if(id === undefined) {
        id = "";
    }
    delete this._savedNodes[id];
}

parsegraph_Caret.prototype.restore = function(id)
{
    if(!this._savedNodes) {
        throw new Error("No saved nodes were found for the provided ID '" + id + "'");
    }
    var loadedNode = this._savedNodes[id];
    if(loadedNode == null) {
        throw new Error("No node found for the provided ID '" + id + "'");
    }
    this._nodes[this._nodes.length - 1] = loadedNode;
}
parsegraph_Caret.prototype.moveTo = parsegraph_Caret.prototype.restore;

parsegraph_Caret.prototype.moveToRoot = function()
{
    this._nodes[this._nodes.length - 1] = this._nodeRoot;
};

parsegraph_Caret.prototype.pop = function()
{
    if(this._nodes.length <= 1) {
        throw parsegraph_createException(parsegraph_NO_NODE_FOUND);
    }
    this._nodes.pop();
};

parsegraph_Caret.prototype.spawnMove = function(inDirection, newContent, newAlignmentMode)
{
    var created = this.spawn(inDirection, newContent, newAlignmentMode);
    this.move(inDirection);
    return created;
};

parsegraph_Caret.prototype.replace = function()
{
    // Retrieve the arguments.
    var node = this.node();
    var withContent = arguments[0];
    if(arguments.length > 1) {
        node = node.nodeAt(parsegraph_readNodeDirection(arguments[0]));
        withContent = arguments[1];
    }

    // Set the node type.
    withContent = parsegraph_readNodeType(withContent);
    node.setType(withContent);
    if(this._mathMode) {
        switch(newType) {
        case parsegraph_BLOCK:
            this.node().setBlockStyle(parsegraph_BLOCK_MATH_STYLE);
            break;
        case parsegraph_SLOT:
            this.node().setBlockStyle(parsegraph_SLOT_MATH_STYLE);
            break;
        }
    }
};

parsegraph_Caret.prototype.at = function(inDirection)
{
    inDirection = parsegraph_readNodeDirection(inDirection);
    if(this.node().hasNode(inDirection)) {
        return this.node().nodeAt(inDirection).type();
    }
};

parsegraph_Caret.prototype.align = function(inDirection, newAlignmentMode)
{
    // Interpret the arguments.
    inDirection = parsegraph_readNodeDirection(inDirection);
    newAlignmentMode = parsegraph_readNodeAlignment(newAlignmentMode);

    this.node().setNodeAlignmentMode(inDirection, newAlignmentMode);
    if(newAlignmentMode != parsegraph_DO_NOT_ALIGN) {
        this.node().setNodeFit(parsegraph_NODE_FIT_EXACT);
    }
};

parsegraph_Caret.prototype.overlapAxis = function(inDirection, newAxisOverlap)
{
    if(arguments.length === 0) {
        this.node().setAxisOverlap(parsegraph_ALLOW_AXIS_OVERLAP);
        return;
    }
    if(arguments.length === 1) {
        this.node().setAxisOverlap(parsegraph_readAxisOverlap(arguments[0]));
        return;
    }
    inDirection = parsegraph_readNodeDirection(inDirection);
    newAxisOverlap = parsegraph_readAxisOverlap(newAxisOverlap);
    this.node().setAxisOverlap(inDirection, newAxisOverlap);
};
parsegraph_Caret.prototype.axisOverlap = parsegraph_Caret.prototype.overlapAxis;

parsegraph_Caret.prototype.pull = function(given)
{
    given = parsegraph_readNodeDirection(given);
    if(this.node().isRoot() || this.node().parentDirection() === parsegraph_OUTWARD) {
        if(parsegraph_isVerticalNodeDirection(given)) {
            this.node().setLayoutPreference(parsegraph_PREFER_VERTICAL_AXIS);
        }
        else {
            this.node().setLayoutPreference(parsegraph_PREFER_HORIZONTAL_AXIS);
        }
        return;
    }
    if(
        parsegraph_getNodeDirectionAxis(given)
        == parsegraph_getNodeDirectionAxis(this.node().parentDirection())
    ) {
        //console.log(parsegraph_nameLayoutPreference(parsegraph_PREFER_PARENT_AXIS));
        this.node().setLayoutPreference(parsegraph_PREFER_PARENT_AXIS);
    }
    else {
        //console.log(parsegraph_nameLayoutPreference(parsegraph_PREFER_PERPENDICULAR_AXIS));
        this.node().setLayoutPreference(parsegraph_PREFER_PERPENDICULAR_AXIS);
    }
};

parsegraph_Caret.prototype.shrink = function()
{
    var node = this.node();
    if(arguments.length > 0) {
        node = node.nodeAt(parsegraph_readNodeDirection(arguments[0]));
    }
    if(node) {
        node.setScale(parsegraph_SHRINK_SCALE);
    }
};

parsegraph_Caret.prototype.grow = function()
{
    var node = this.node();
    if(arguments.length > 0) {
        node = node.nodeAt(parsegraph_readNodeDirection(arguments[0]));
    }
    if(node) {
        node.setScale(1.0);
    }
};

parsegraph_Caret.prototype.fitExact = function()
{
    var node = this.node();
    if(arguments.length > 0) {
        node = node.nodeAt(parsegraph_readNodeDirection(arguments[0]));
    }
    node.setNodeFit(parsegraph_NODE_FIT_EXACT);
};

parsegraph_Caret.prototype.fitLoose = function()
{
    var node = this.node();
    if(arguments.length > 0) {
        node = node.nodeAt(parsegraph_readNodeDirection(arguments[0]));
    }
    node.setNodeFit(parsegraph_NODE_FIT_LOOSE);
};

parsegraph_Caret.prototype.fitNaive = function()
{
    var node = this.node();
    if(arguments.length > 0) {
        node = node.nodeAt(parsegraph_readNodeDirection(arguments[0]));
    }
    node.setNodeFit(parsegraph_NODE_FIT_NAIVE);
};

parsegraph_Caret.prototype.label = function(/* ... */)
{
    var node, text, font;
    switch(arguments.length) {
    case 0:
        return this.node().label();
    case 1:
        node = this.node();
        text = arguments[0];
        font = this.font();
        break;
    case 2:
        if(typeof arguments[1] === "object") {
            node = this.node();
            text = arguments[0];
            font = arguments[1];
        }
        else {
            node = node.nodeAt(parsegraph_readNodeDirection(arguments[0]));
            text = arguments[1];
            font = this.font();
            //console.log(typeof arguments[0]);
            //console.log(typeof arguments[1]);
        }
        break;
    case 3:
        node = node.nodeAt(parsegraph_readNodeDirection(arguments[0]));
        text = arguments[1];
        font = arguments[2];
        break;
    }
    node.setLabel(text, font);
};

parsegraph_Caret.prototype.select = function()
{
    var node = this.node();
    if(arguments.length > 0) {
        node = node.nodeAt(parsegraph_readNodeDirection(arguments[0]));
    }
    node.setSelected(true);
};

parsegraph_Caret.prototype.selected = function()
{
    var node = this.node();
    if(arguments.length > 0) {
        node = node.nodeAt(parsegraph_readNodeDirection(arguments[0]));
    }
    return node.isSelected();
};

parsegraph_Caret.prototype.deselect = function()
{
    var node = this.node();
    if(arguments.length > 0) {
        node = node.nodeAt(parsegraph_readNodeDirection(arguments[0]));
    }
    node.setSelected(false);
};

parsegraph_Caret.prototype.graph = function()
{
    return this.node().graph();
};

/**
 * Returns the initiall provided node.
 */
parsegraph_Caret.prototype.root = function()
{
    return this._nodeRoot;
};

File: src/graph/Caret.c
#include "Caret.h"
#include "initialize.h"
#include "LayoutPreference.h"
#include "id.h"
#include "NodeFit.h"
#include "Surface.h"
#include "../die.h"
#include "Status.h"
#include "NodeAlignment.h"

parsegraph_Caret* parsegraph_Caret_new(parsegraph_Surface* surface, parsegraph_Node* nodeRoot)
{
    apr_pool_t* pool = 0;
    if(!nodeRoot) {
        if(APR_SUCCESS != apr_pool_create(&pool, surface->pool)) {
            parsegraph_die("Failed to create APR pool for parsegraph_Caret");
        }
        nodeRoot = parsegraph_Node_new(pool, parsegraph_DEFAULT_NODE_TYPE, 0, 0);
    }
    else {
        if(APR_SUCCESS != apr_pool_create(&pool, nodeRoot->pool)) {
            parsegraph_die("Failed to create APR pool for parsegraph_Caret");
        }
    }
    parsegraph_Caret* caret = apr_palloc(pool, sizeof(*caret));
    caret->surface = surface;
    caret->pool = pool;

    caret->_mathMode = 0;

    caret->_nodeRoot = nodeRoot;

    // Stack of nodes.
    caret->_nodes = parsegraph_ArrayList_new(pool);
    if(caret->_nodeRoot) {
        parsegraph_ArrayList_push(caret->_nodes, caret->_nodeRoot);
    }

    // A mapping of nodes to their saved names.
    caret->_savedNodes = 0;

    caret->_labels = parsegraph_ArrayList_new(pool);

    caret->_glyphAtlas = 0;
    return caret;
}

void parsegraph_Caret_destroy(parsegraph_Caret* caret)
{
    parsegraph_ArrayList_destroy(caret->_nodes);
    if(caret->_savedNodes) {
        parsegraph_ArrayList_destroy(caret->_savedNodes);
    }
    parsegraph_ArrayList_destroy(caret->_labels);
    apr_pool_destroy(caret->pool);
}

void parsegraph_Caret_setMathMode(parsegraph_Caret* caret, int mathMode)
{
    caret->_mathMode = mathMode;
    parsegraph_Node* curr = parsegraph_Caret_node(caret);
    if(mathMode) {
        switch(parsegraph_Node_type(curr)) {
        case parsegraph_BLOCK:
            parsegraph_Node_setBlockStyle(curr, parsegraph_BLOCK_MATH_STYLE);
            break;
        case parsegraph_SLOT:
            parsegraph_Node_setBlockStyle(curr, parsegraph_SLOT_MATH_STYLE);
            break;
        }
    }
}

int parsegraph_Caret_mathMode(parsegraph_Caret* caret)
{
    return caret->_mathMode;
}

void parsegraph_Caret_setGlyphAtlas(parsegraph_Caret* caret, parsegraph_GlyphAtlas* glyphAtlas)
{
    caret->_glyphAtlas = glyphAtlas;
}

parsegraph_GlyphAtlas* parsegraph_Caret_glyphAtlas(parsegraph_Caret* caret)
{
    if(!caret->_glyphAtlas) {
        parsegraph_die("Caret does not have a GlyphAtlas");
    }
    return caret->_glyphAtlas;
}

parsegraph_Node* parsegraph_Caret_node(parsegraph_Caret* caret)
{
    if(parsegraph_ArrayList_length(caret->_nodes) == 0) {
        parsegraph_abort(parsegraph_NO_NODE_FOUND);
    }
    return parsegraph_ArrayList_at(caret->_nodes, parsegraph_ArrayList_length(caret->_nodes) - 1);
}

int parsegraph_Caret_has(parsegraph_Caret* caret, const char* inDirection)
{
    return parsegraph_Node_hasNode(parsegraph_Caret_node(caret), parsegraph_readNodeDirection(inDirection));
}

void parsegraph_Caret_spawn(parsegraph_Caret* caret, const char* inDirection, const char* newType, const char* newAlignmentMode)
{
    // Interpret the given direction and type for ease-of-use.
    int inDirectionVal = parsegraph_readNodeDirection(inDirection);
    int newTypeVal = parsegraph_readNodeType(newType);

    // Spawn a node in the given direction.
    parsegraph_Node* created = parsegraph_Node_spawnNode(parsegraph_Caret_node(caret), inDirectionVal, newTypeVal);

    // Use the given alignment mode.
    if(newAlignmentMode) {
        int newAlignmentModeVal = parsegraph_readNodeAlignment(newAlignmentMode);
        parsegraph_Caret_align(caret, inDirection, newAlignmentMode);
        if(newAlignmentModeVal != parsegraph_DO_NOT_ALIGN) {
            parsegraph_Node_setNodeFit(parsegraph_Caret_node(caret), parsegraph_NODE_FIT_EXACT);
        }
    }

    if(caret->_mathMode) {
        switch(newTypeVal) {
        case parsegraph_BLOCK:
            parsegraph_Node_setBlockStyle(created, parsegraph_BLOCK_MATH_STYLE);
            break;
        case parsegraph_SLOT:
            parsegraph_Node_setBlockStyle(created, parsegraph_SLOT_MATH_STYLE);
            break;
        }
    }

    parsegraph_Node_unref(created);
};

parsegraph_Node* parsegraph_Caret_connect(parsegraph_Caret* caret, const char* inDirection, parsegraph_Node* node)
{
    // Interpret the given direction for ease-of-use.
    int inDirectionVal = parsegraph_readNodeDirection(inDirection);

    parsegraph_Node_connectNode(parsegraph_Caret_node(caret), inDirectionVal, node);

    return node;
};

parsegraph_Node* parsegraph_Caret_disconnect(parsegraph_Caret* caret, const char* inDirection)
{
    if(inDirection) {
        // Interpret the given direction for ease-of-use.
        int inDirectionVal = parsegraph_readNodeDirection(inDirection);
        return parsegraph_Node_disconnectNode(parsegraph_Caret_node(caret), inDirectionVal);
    }

    if(parsegraph_Node_isRoot(parsegraph_Caret_node(caret))) {
        parsegraph_die("A root node cannot be disconnected.");
    }

    return parsegraph_Node_disconnectNode(parsegraph_Node_parentNode(parsegraph_Caret_node(caret)), parsegraph_reverseNodeDirection(parsegraph_Node_parentDirection(parsegraph_Caret_node(caret))));
}

void parsegraph_Caret_crease(parsegraph_Caret* caret, const char* inDirection)
{
    // Interpret the given direction for ease-of-use.
    parsegraph_Node* node;
    if(!inDirection) {
        node = parsegraph_Caret_node(caret);
    }
    else {
        int inDirectionVal = parsegraph_readNodeDirection(inDirection);
        node = parsegraph_Node_nodeAt(parsegraph_Caret_node(caret), inDirectionVal);
    }

    // Create a new paint group for the connection.
    if(!parsegraph_Node_localPaintGroup(node)) {
        parsegraph_Node_setPaintGroup(node, 1);
    }
}

void parsegraph_Caret_uncrease(parsegraph_Caret* caret, const char* inDirection)
{
    // Interpret the given direction for ease-of-use.
    parsegraph_Node* node;
    if(!inDirection) {
        node = parsegraph_Caret_node(caret);
    }
    else {
        int inDirectionVal = parsegraph_readNodeDirection(inDirection);
        node = parsegraph_Node_nodeAt(parsegraph_Caret_node(caret), inDirectionVal);
    }

    // Remove the paint group.
    parsegraph_Node_setPaintGroup(node, 0);
}

int parsegraph_Caret_isCreased(parsegraph_Caret* caret, const char* inDirection)
{
    // Interpret the given direction for ease-of-use.
    int readDirection = parsegraph_readNodeDirection(inDirection);

    parsegraph_Node* node;
    if(!inDirection) {
        node = parsegraph_Caret_node(caret);
    }
    else {
        node = parsegraph_Node_nodeAt(parsegraph_Caret_node(caret), readDirection);
    }

    return parsegraph_Node_localPaintGroup(node);
}

int parsegraph_Caret_creased(parsegraph_Caret* caret, const char* inDirection)
{
    return parsegraph_Caret_isCreased(caret, inDirection);
}

void parsegraph_Caret_erase(parsegraph_Caret* caret, const char* inDirection)
{
    int inDirectionVal = parsegraph_readNodeDirection(inDirection);
    parsegraph_Node_eraseNode(parsegraph_Caret_node(caret), inDirectionVal);
};

void parsegraph_Caret_onClick(parsegraph_Caret* caret, int(*clickListener)(parsegraph_Node*, const char*, void*), void* thisArg)
{
    parsegraph_Node_setClickListener(parsegraph_Caret_node(caret), clickListener, thisArg);
}

void parsegraph_Caret_onChange(parsegraph_Caret* caret, void(*changeListener)(void*, parsegraph_Node*), void* thisArg)
{
    parsegraph_Node_setChangeListener(parsegraph_Caret_node(caret), changeListener, thisArg);
}

void parsegraph_Caret_onKey(parsegraph_Caret* caret, int(*keyListener)(parsegraph_Node*, const char*, void*), void* thisArg)
{
    parsegraph_Node_setKeyListener(parsegraph_Caret_node(caret), keyListener, thisArg);
}

void parsegraph_Caret_move(parsegraph_Caret* caret, const char* toDirection)
{
    int toDirectionVal = parsegraph_readNodeDirection(toDirection);
    parsegraph_Node* dest = parsegraph_Node_nodeAt(parsegraph_Caret_node(caret), toDirectionVal);
    if(!dest) {
        parsegraph_abort(parsegraph_NO_NODE_FOUND);
    }
    parsegraph_ArrayList_replace(caret->_nodes, parsegraph_ArrayList_length(caret->_nodes) -1, dest);
}

void parsegraph_Caret_push(parsegraph_Caret* caret)
{
    parsegraph_ArrayList_push(caret->_nodes, parsegraph_Caret_node(caret));
}

static parsegraph_SavedCaret* parsegraph_Caret_find(parsegraph_Caret* caret, const char* id)
{
    for(int i = 0; i < parsegraph_ArrayList_length(caret->_savedNodes); ++i) {
        parsegraph_SavedCaret* sc = parsegraph_ArrayList_at(caret->_savedNodes, i);
        if(!strcmp(sc->id, id)) {
            return sc;
        }
    }
    return 0;
}

const char* parsegraph_Caret_save(parsegraph_Caret* caret, const char* id)
{
    if(id && strlen(id) > 254) {
        parsegraph_die("Caret ID is too large");
    }

    if(!caret->_savedNodes) {
        caret->_savedNodes = parsegraph_ArrayList_new(caret->pool);
    }

    parsegraph_SavedCaret* sc;
    char buf[255];
    memset(buf, 0, sizeof(char)*255);
    if(!id) {
        parsegraph_generateID(buf, 255, "Caret");
    }
    else {
        sc = parsegraph_Caret_find(caret, id);
        if(sc) {
            sc->node = parsegraph_Caret_node(caret);
            return sc->id;
        }
    }

    sc = apr_palloc(caret->pool, sizeof(*sc));
    strncpy(sc->id, id ? id : buf, 255);
    sc->node = parsegraph_Caret_node(caret);
    parsegraph_ArrayList_push(caret->_savedNodes, sc);
    return sc->id;
}

void parsegraph_Caret_clearSave(parsegraph_Caret* caret, const char* id)
{
    if(!caret->_savedNodes) {
        return;
    }
    for(int i = 0; i < parsegraph_ArrayList_length(caret->_savedNodes); ++i) {
        parsegraph_SavedCaret* sc = parsegraph_ArrayList_at(caret->_savedNodes, i);
        if(!strcmp(sc->id, id)) {
            parsegraph_ArrayList_splice(caret->_savedNodes, i, 1);
            return;
        }
    }
}

void parsegraph_Caret_restore(parsegraph_Caret* caret, const char* id)
{
    if(!caret->_savedNodes) {
        parsegraph_die("No saved nodes were found for the provided ID '%s'", id);
    }

    parsegraph_SavedCaret* sc = parsegraph_Caret_find(caret, id);
    if(!sc) {
        parsegraph_die("No node found for the provided ID '%s'", id);
    }
    parsegraph_Node* loadedNode = sc->node;
    parsegraph_ArrayList_replace(caret->_nodes, parsegraph_ArrayList_length(caret->_nodes) - 1, loadedNode);
}

void parsegraph_Caret_moveTo(parsegraph_Caret* caret, const char* id)
{
    parsegraph_Caret_restore(caret, id);
}

void parsegraph_Caret_moveToRoot(parsegraph_Caret* caret)
{
    parsegraph_ArrayList_replace(caret->_nodes, parsegraph_ArrayList_length(caret->_nodes) - 1, caret->_nodeRoot);
}

void parsegraph_Caret_pop(parsegraph_Caret* caret)
{
    if(parsegraph_ArrayList_length(caret->_nodes) <= 1) {
        parsegraph_abort(parsegraph_NO_NODE_FOUND);
    }
    parsegraph_ArrayList_pop(caret->_nodes);
};

parsegraph_Node* parsegraph_Caret_spawnMove(parsegraph_Caret* caret, const char* inDirection, const char* newContent, const char* newAlignmentMode)
{
    parsegraph_Caret_spawn(caret, inDirection, newContent, newAlignmentMode);
    parsegraph_Caret_move(caret, inDirection);
    return parsegraph_Caret_node(caret);
}

void parsegraph_Caret_replace(parsegraph_Caret* caret, const char* arg1, const char* arg2)
{
    // Retrieve the arguments.
    parsegraph_Node* node = parsegraph_Caret_node(caret);
    const char* withContent = arg1;
    if(arg2) {
        node = parsegraph_Node_nodeAt(node, parsegraph_readNodeDirection(arg1));
        withContent = arg2;
    }

    // Set the node type.
    int withContentVal = parsegraph_readNodeType(withContent);

    parsegraph_Node_setType(node, withContentVal);
    if(caret->_mathMode) {
        switch(withContentVal) {
        case parsegraph_BLOCK:
            parsegraph_Node_setBlockStyle(node, parsegraph_BLOCK_MATH_STYLE);
            break;
        case parsegraph_SLOT:
            parsegraph_Node_setBlockStyle(node, parsegraph_SLOT_MATH_STYLE);
            break;
        }
    }
}

int parsegraph_Caret_at(parsegraph_Caret* caret, const char* inDirection)
{
    int inDirectionVal = parsegraph_readNodeDirection(inDirection);
    if(parsegraph_Node_hasNode(parsegraph_Caret_node(caret), inDirectionVal)) {
        return parsegraph_Node_type(parsegraph_Node_nodeAt(parsegraph_Caret_node(caret), inDirectionVal));
    }
    parsegraph_abort(parsegraph_NO_NODE_FOUND);
};

void parsegraph_Caret_align(parsegraph_Caret* caret, const char* inDirection, const char* newAlignmentMode)
{
    // Interpret the arguments.
    int inDirectionVal = parsegraph_readNodeDirection(inDirection);
    int newAlignmentModeVal = parsegraph_readNodeAlignment(newAlignmentMode);

    parsegraph_Node_setNodeAlignmentMode(parsegraph_Caret_node(caret), inDirectionVal, newAlignmentModeVal);
    if(newAlignmentModeVal != parsegraph_DO_NOT_ALIGN) {
        parsegraph_Node_setNodeFit(parsegraph_Caret_node(caret), parsegraph_NODE_FIT_EXACT);
    }
}

void parsegraph_Caret_pull(parsegraph_Caret* caret, const char* given)
{
    int givenVal = parsegraph_readNodeDirection(given);
    if(parsegraph_Node_isRoot(parsegraph_Caret_node(caret)) || parsegraph_Node_parentDirection(parsegraph_Caret_node(caret)) == parsegraph_OUTWARD) {
        if(parsegraph_isVerticalNodeDirection(givenVal)) {
            parsegraph_Node_setLayoutPreference(parsegraph_Caret_node(caret), parsegraph_PREFER_VERTICAL_AXIS);
        }
        else {
            parsegraph_Node_setLayoutPreference(parsegraph_Caret_node(caret), parsegraph_PREFER_HORIZONTAL_AXIS);
        }
        return;
    }
    if(
        parsegraph_getNodeDirectionAxis(givenVal)
        == parsegraph_getNodeDirectionAxis(parsegraph_Node_parentDirection(parsegraph_Caret_node(caret)))
    ) {
        parsegraph_Node_setLayoutPreference(parsegraph_Caret_node(caret), parsegraph_PREFER_PARENT_AXIS);
    }
    else {
        parsegraph_Node_setLayoutPreference(parsegraph_Caret_node(caret), parsegraph_PREFER_PERPENDICULAR_AXIS);
    }
}

void parsegraph_Caret_shrink(parsegraph_Caret* caret, const char* inDir)
{
    parsegraph_Node* node = parsegraph_Caret_node(caret);
    if(inDir) {
        node = parsegraph_Node_nodeAt(node, parsegraph_readNodeDirection(inDir));
    }
    if(node) {
        parsegraph_Node_setScale(node, parsegraph_SHRINK_SCALE);
    }
};

void parsegraph_Caret_grow(parsegraph_Caret* caret, const char* inDir)
{
    parsegraph_Node* node = parsegraph_Caret_node(caret);
    if(inDir) {
        node = parsegraph_Node_nodeAt(node, parsegraph_readNodeDirection(inDir));
    }
    if(node) {
        parsegraph_Node_setScale(node, 1.0);
    }
};

void parsegraph_Caret_fitExact(parsegraph_Caret* caret, const char* inDir)
{
    parsegraph_Node* node = parsegraph_Caret_node(caret);
    if(inDir) {
        node = parsegraph_Node_nodeAt(node, parsegraph_readNodeDirection(inDir));
    }
    parsegraph_Node_setNodeFit(node, parsegraph_NODE_FIT_EXACT);
};

void parsegraph_Caret_fitLoose(parsegraph_Caret* caret, const char* inDir)
{
    parsegraph_Node* node = parsegraph_Caret_node(caret);
    if(inDir) {
        node = parsegraph_Node_nodeAt(node, parsegraph_readNodeDirection(inDir));
    }
    parsegraph_Node_setNodeFit(node, parsegraph_NODE_FIT_LOOSE);
};

void parsegraph_Caret_label(parsegraph_Caret* caret, const char* text, parsegraph_GlyphAtlas* glyphAtlas, const char* inDir)
{
    parsegraph_Node* node = parsegraph_Caret_node(caret);
    if(inDir) {
        node = parsegraph_Node_nodeAt(node, parsegraph_readNodeDirection(inDir));
    }
    if(!glyphAtlas) {
        glyphAtlas = parsegraph_Caret_glyphAtlas(caret);
    }
    parsegraph_Node_setLabelUTF8(node, text, strlen(text), glyphAtlas);
}

void parsegraph_Caret_select(parsegraph_Caret* caret, const char* inDir)
{
    parsegraph_Node* node = parsegraph_Caret_node(caret);
    if(inDir) {
        node = parsegraph_Node_nodeAt(node, parsegraph_readNodeDirection(inDir));
    }
    parsegraph_Node_setSelected(node, 1);
};

int parsegraph_Caret_selected(parsegraph_Caret* caret, const char* inDir)
{
    parsegraph_Node* node = parsegraph_Caret_node(caret);
    if(inDir) {
        node = parsegraph_Node_nodeAt(node, parsegraph_readNodeDirection(inDir));
    }
    return parsegraph_Node_isSelected(node);
};

void parsegraph_Caret_deselect(parsegraph_Caret* caret, const char* inDir)
{
    parsegraph_Node* node = parsegraph_Caret_node(caret);
    if(inDir) {
        node = parsegraph_Node_nodeAt(node, parsegraph_readNodeDirection(inDir));
    }
    parsegraph_Node_setSelected(node, 0);
};

parsegraph_Node* parsegraph_Caret_root(parsegraph_Caret* caret)
{
    return caret->_nodeRoot;
};
